util/gatewaycrafter/pullpingmap.go

```diff
@@ -0,0 +1,38 @@
+package gatewaycrafter
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/google/uuid"
+)
+
+// PullPingMap is a thread-save map used for coordinating pullpings
+type PullPingMap struct {
+	m           sync.Mutex
+	localImages map[string]string
+}
+
+// NewPullPingMap returns a new PullPingMap
+func NewPullPingMap() *PullPingMap {
+	return &PullPingMap{
+		localImages: map[string]string{},
+	}
+}
+
+// Get fetches an existing entry from the map or returns false if none exists
+func (ppm *PullPingMap) Get(k string) (string, bool) {
+	ppm.m.Lock()
+	defer ppm.m.Unlock()
+	v, ok := ppm.localImages[k]
+	return v, ok
+}
+
+// Insert creates a new entry for the value under sessionID/<uuid>
+func (ppm *PullPingMap) Insert(sessionID, v string) string {
+	k := fmt.Sprintf("sess-%s/pullping:%s", sessionID, uuid.New().String())
```

Comment:
I wonder if we should make this something a little more transparent than UUID. It could help with debuggability.

Reply:
possibly, but we need to ensure it's unique.

Adding a reference to the WAIT/END block (e.g. Earthfile and line number), plus an index alone wouldn't guarentee uniqueness since it could be executed multiple times with different build-args.

Any thoughts on an easy-to-implement solution?

Reply:
Maybe it can be based on the name of the final image, but concatenated with some uniqueness, like the uuid, or an incrementing index.